src/compiler/moduleNameResolver.ts

```diff
@@ -354,7 +354,10 @@ namespace ts {
         }
         const conditions = features & NodeResolutionFeatures.Exports ? features & NodeResolutionFeatures.EsmMode ? ["node", "import", "types"] : ["node", "require", "types"] : [];
         const moduleResolutionState: ModuleResolutionState = { compilerOptions: options, host, traceEnabled, failedLookupLocations, packageJsonInfoCache: cache, features, conditions };
-        let resolved = primaryLookup();
+        const loader: ResolutionKindSpecificLoader = (extensions, candidate, onlyRecordFailures, state) => nodeLoadModuleByRelativeName(extensions, candidate, onlyRecordFailures, state, /*considerPackageJson*/ true);
+        const optional = tryLoadModuleUsingOptionalResolutionSettings(Extensions.DtsOnly, typeReferenceDirectiveName, containingDirectory || "", loader, moduleResolutionState);

```

Comment:
Well paths uses baseUrl.. And then `tryLoadModuleUsingOptionalResolutionSettings` has some baseUrl and rootDirs logic as well.. wouldnt we want typeRoots instead of baseUrl and probably rootDirs as well for type reference directive

Reply:
I don't think that's the case? Generally, the model we all have is that `///<reference types=` resolves exactly like an `import` statement does. An import statement uses the full `tryLoadModuleUsingOptionalResolutionSettings` function, so I think we also should use it here.

(Plus, `typeRoots` is already explicitly handled when looking up `@types` names inside `primaryLookup`)

Reply:
I think the only 🤷‍♀️❓🤷‍♀️ kinda thing is weather a `typeRoots` resolution should have priority over a `paths`/`baseUrl` one or not. I'd argue that `paths` and `baseUrl` are more specific compiler options, so should have priority (hence this placement).

Reply:
But typeRoots also is in compilerOption and i think is set when user prefers to resolve type reference directive from certain location.. When not set i agree that it baseUrl makes sense

Reply:
Yeah, but `typeRoots` a list of potential directories, while `paths` is exact matches and `baseUrl` is a single directory. That's why `paths` > `baseUrl` > `typeRoots` makes sense to me. (`typeRoots` will never conflict with `rootDirs` as it were because `typeRoots` are only relevant for non-relative names, while `rootDirs` are only relevant for relative names)